Implementing routing [resolves #116]

lhaatveit opened the pull request:
#### Issue description

Implements the routing feature as discussed in #116. 

* Added LoginExchangeResult in LoginFlow which can either indicate a successful connection, or routing to a different server
* Upon being routed, close the current client and try again using the new server

Note that the changes are not handling changes in certificate host name in any way, I'm not sure how you'd like this to be handled.

#### New Public APIs

N/A

#### Additional context

Spec:

* Token: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/2b3eb7e5-d43d-4d1b-bf4d-76b9e3afc791
* Semantics: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/52851226-1211-46c8-ac50-6b7facd08fd9

pivotal-issuemaster commented:
@lhaatveit Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=r2dbc/r2dbc-mssql&pullRequestId=137)!

[Click here](https://cla.pivotal.io/sync/reactor?repositoryId=r2dbc/r2dbc-mssql&pullRequestId=137) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@lhaatveit Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=r2dbc/r2dbc-mssql&pullRequestId=137)!

mp911de commented on src/main/java/io/r2dbc/mssql/message/tds/Decode.java:
```diff
@@ -224,4 +224,35 @@ private static String decodeUnicode(ByteBuf buffer, int length) {
+     * @param buffer the data buffer
+     * @return the decoded {@link RoutingData}
+     */
+    public static RoutingData decodeRoute(ByteBuf buffer) {
```
Let's move decoding into `RoutingData` itself (`public static RoutingData decode(ByteBuf buffer) {`). `Decode` is intended for low-level primitives.

lhaatveit replied:
Done

mp911de commented on src/main/java/io/r2dbc/mssql/message/token/EnvChangeToken.java:
```diff
@@ -69,13 +69,26 @@ public static EnvChangeToken decode(ByteBuf buffer) {
-        int oldValueLen = envChangeType.toByteLength(Decode.asByte(buffer));
-        byte[] oldValue = new byte[oldValueLen];
-        buffer.readBytes(oldValue);
+        if (envChangeType == EnvChangeType.Routing) {
```
We should probably leave a comment for both branches saying that `Routing` has a structured content and other tokens just report old/new values.

lhaatveit replied:
Yes. I have left a comment there now.

mp911de commented on src/main/java/io/r2dbc/mssql/MssqlConnectionFactory.java:
```diff
@@ -50,28 +52,68 @@
+
+                            assert result.getAlternateServerName() != null;
+
+                            MssqlConnectionConfiguration routeConfiguration = configuration.toBuilder()
```
Instead using a builder, we could simply have a `withServer(…)` method that creates a new `MssqlConnectionConfiguration` object. Since both types are in the same package, we don't need to expose this method publicly.

lhaatveit replied:
Done. I have left the `toBuilder` as a private method for now. Let me know if you prefer to have it gone altogether.

mp911de replied:
There's no need for a `toBuilder` method, so let's remove it. These kind of methods are error prone for adding additional properties as they get easily forgotten. Extending a constructor typically causes compile errors if a caller wasn't updated.

mp911de commented on src/main/java/io/r2dbc/mssql/MssqlConnectionConfiguration.java:
```diff
@@ -100,6 +100,38 @@ public static Builder builder() {
+     *
+     * @return a {@link Builder} with this configuration
+     */
+    public Builder toBuilder() {
```
Instead of having a builder method, we could expose a package-private method `MssqlConnectionConfiguration withServer(…)` that returns just a new instance of `MssqlConnectionConfiguration` with hostname/port applied.

mp911de commented on src/main/java/io/r2dbc/mssql/LoginFlow.java:
```diff
@@ -99,8 +101,27 @@ private LoginFlow() {
+                        return;
+                    }
+
+                    if (message instanceof EnvChangeToken) {
```
We keep most items that are provided via `EnvChangeToken` within the client (e.g. transaction id, database, locale). How about moving `RoutingData` into `Client` and expose it as `Optional<RoutingData> getRedirect()`?

lhaatveit replied:
I have reverted the changes in {{LoginFlow}} and added a listener on the client for the redirect instead. I think this is more in line with the existing pattern. Let me know what you think.

mp911de commented on src/main/java/io/r2dbc/mssql/client/LoginExchangeResult.java:
```diff
@@ -0,0 +1,86 @@
+    private final Outcome outcome;
+
+    @Nullable
+    private final String alternateServerName;
```
How about reusing `RoutingData` instead of having two fields?

mp911de commented on src/main/java/io/r2dbc/mssql/message/tds/RoutingData.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Represents an alternative route.
+ */
+public final class RoutingData {
```
How about naming this type `Redirect`, `Server` or `Endpoint`? `RoutingData` sounds a bit clunky.

lhaatveit replied:
I picked {{RoutingData}} to use the same term as in the spec, but I agree. Renamed to {{Redirect}}.

lhaatveit commented:
@mp911de I removed the unit test for the login flow since I reverted those changes. I am unsure how I should proceed with creating tests for the {{ReactorNettyClient}} and/or {{MssqlConnectionFactory}}.

mp911de commented on src/main/java/io/r2dbc/mssql/client/ReactorNettyClient.java:
```diff
@@ -778,6 +789,19 @@ public void onEnvironmentChange(EnvironmentChangeEvent event) {
+
+            if (event.getToken().getChangeType() == EnvChangeToken.EnvChangeType.Routing) {
+
+                Redirect redirect = Redirect.decode(Unpooled.wrappedBuffer(event.getToken().getNewValue()));
```
👍

lhaatveit commented:
I just noticed that the SQL server has stopped sending the routing message today (no idea why), so the latest changes have not actually been tested properly. Let me find a way to verify that they actually work before we merge this.

lhaatveit commented:
Okay, got it tested now. Working fine.

mp911de commented:
I pushed meanwhile the code onto the [`routing` branch](https://github.com/r2dbc/r2dbc-mssql/tree/routing).

What configuration is required on Azure so that the server sends a routing token?

lhaatveit commented:
The way I've been testing is by accessing Azure SQL through an AKS cluster. I would guess it can also be done by enabling replication, but I haven't tried that yet...

mp911de commented:
I found something: It's setting the connection type policy via `az sql server conn-policy update Redirect`. I found a missing bit regarding `hostNameInCertificate`: The JDBC driver updates the trust setting if the redirect host falls is trusted by the `hostNameInCertificate` expression. I added the same functionality to the R2DBC driver for the same experience as with JDBC.

mp911de commented:
Thank you for your contribution. That's merged, polished, and backported now.
